utils/zerovec/src/map/vecs.rs

```diff
@@ -13,19 +14,49 @@ use core::mem;
 
 /// Trait abstracting over [`ZeroVec`] and [`VarZeroVec`], for use in [`ZeroMap`](super::ZeroMap). You
 /// should not be implementing or calling this trait directly.
-pub trait ZeroVecLike<'a, T: ?Sized> {
+pub trait BorrowedZeroVecLike<'a, T: ?Sized> {
     /// The type received by `Self::binary_search()`
     type NeedleType: ?Sized;
     /// The type returned by `Self::get()`
-    type GetType: ?Sized;
-    /// The type returned by `Self::remove()` and `Self::replace()`
-    type OwnedType;
+    type GetType: ?Sized + 'static;
     /// Search for a key in a sorted vector, returns `Ok(index)` if found,
     /// returns `Err(insert_index)` if not found, where `insert_index` is the
     /// index where it should be inserted to maintain sort order.
     fn binary_search(&self, k: &Self::NeedleType) -> Result<usize, usize>;
     /// Get element at `index`
     fn get(&self, index: usize) -> Option<&Self::GetType>;
+    /// The length of this vector
+    fn len(&self) -> usize;
+    /// Check if this vector is in ascending order according to `T`s `Ord` impl
+    fn is_ascending(&self) -> bool;
+    /// Check if this vector is empty
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Trait abstracting over [`ZeroVec`] and [`VarZeroVec`], for use in [`ZeroMap`](super::ZeroMap). You
+/// should not be implementing or calling this trait directly.
+///
+/// This trait augments [`BorrowedZeroVecLike`] with methods allowing for taking
+/// longer references to the underlying buffer, for borrowed-only vector types.
+pub trait BorrowedOnlyZeroVecLike<'a, T: ?Sized>: BorrowedZeroVecLike<'a, T> {
+    /// Get element at `index`, with a longer lifetime
+    fn get_lengthened(&self, index: usize) -> Option<&'a Self::GetType>;
+}
+
+/// Trait abstracting over [`ZeroVec`] and [`VarZeroVec`], for use in [`ZeroMap`](super::ZeroMap). You
+/// should not be implementing or calling this trait directly.
+///
+/// This trait augments [`BorrowedZeroVecLike`] with methods allowing for mutation of the underlying
+/// vector for owned vector types.
+pub trait ZeroVecLike<'a, T: ?Sized>: BorrowedZeroVecLike<'a, T> {
+    /// The type returned by `Self::remove()` and `Self::replace()`
+    type OwnedType;
+    /// A fully borrowed version of this
+    type BorrowedVersion: BorrowedZeroVecLike<'a, T, NeedleType = Self::NeedleType, GetType = Self::GetType>
```

Comment:
Nit: I don't like "version" in the name, because "version" is overloaded.  Suggestions:

- `Borrowed`
- `BorrowedType`
- `BorrowedLike`
- `BorrowedZeroVecLike`

Reply:
How about BorrowedVariant?